qualtran/bloqs/arithmetic/comparison.py

```diff
@@ -465,50 +467,172 @@ def _has_unitary_(self):
 
     @property
     def signature(self):
-        return Signature.build(a=self.a_bitsize, b=self.b_bitsize, target=1)
-
-    def short_name(self) -> str:
-        return "a>b"
-
-    def t_complexity(self) -> 'TComplexity':
-        return t_complexity(LessThanEqual(self.a_bitsize, self.b_bitsize))
+        return Signature(
+            [
+                Register('a', bitsize=self.bitsize),
+                Register('b', bitsize=self.bitsize),
+                Register('target', bitsize=1),
+            ]
+        )
 
-    def wire_symbol(self, soq: Soquet) -> WireSymbol:
-        if soq.reg.name == 'a':
-            return TextBox("In(a)")
-        if soq.reg.name == 'b':
-            return TextBox("In(b)")
-        elif soq.reg.name == 'target':
-            return TextBox("⨁(a > b)")
+    def on_classical_vals(
+        self, a: 'ClassicalValT', b: 'ClassicalValT', target: 'ClassicalValT'
+    ) -> Dict[str, 'ClassicalValT']:
+        # target is a 1-bit register so we assert that it's classical value is binary.
+        assert target == (target % 2)
+
+        if b > a:
+            target = (target + 1) % 2
+
+        return {'a': a, 'b': b, 'target': target}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', a: SoquetT, b: SoquetT, target: SoquetT
+    ) -> Dict[str, 'SoquetT']:
+
+        # Base Case: Comparing two qubits.
+        # Signed doesn't matter because we can't represent signed integers with 1 qubit.
+        if self.bitsize == 1:
+            # We use a specially controlled Toffolli gate to implement GreaterThan.
+            # If a is 0 and b is 1 then b > a and we can flip the target bit.
+            ctrls = [a, b]
+            ctrls, target = bb.add(MultiControlX(cvs=(0, 1)), ctrls=ctrls, x=target)
+            a, b = ctrls
+            # Return the output registers.
+            return {'a': a, 'b': b, 'target': target}
+
+        # Allocate lists to store ancillas generated by the logical-and and control pairs input
+        # into logical-ands.
+        ancillas = []
+        and_ctrls = []
+
+        # If the input registers are unsigned we need to append a sign bit to them in order to use
+        # the 1's complement trick.
+        if not self.signed:
+            a_sign = bb.allocate(n=1)
+            a_split = bb.split(a)
+            a = bb.join(np.concatenate([[a_sign], a_split]))
+
+            b_sign = bb.allocate(n=1)
+            b_split = bb.split(b)
+            b = bb.join(np.concatenate([[b_sign], b_split]))
+
+        # Create variable true_bitsize to account for sign bit in bloq construction.
+        true_bitsize = self.bitsize if self.signed else (self.bitsize + 1)
+
+        # Flip all the bits in the x register.
+        a_split = bb.split(a)
+
+        for i in range(true_bitsize):
+            a_split[i] = bb.add(XGate(), q=a_split[i])
+        b_split = bb.split(b)
+
+        # Iteratively implements the left adder circuit building block of the Gidney Adder. On
```

Comment:
you don't need to implement the decompisition of the adder gate. We already have the Gidney adder in qualtran `qualtran.bloqs.arithmetic.addition.Add` which you can use in your decomposition.

https://github.com/quantumlib/Qualtran/blob/642dec21cb6844bbe2f5f33e4312a4951f73e8c9/qualtran/bloqs/arithmetic/addition.py#L48

Reply:
I implemented the decomposition because we only need to see the most significant bit of the b register after the add and then use that to modify the target qubit. After this we need to uncompute the add so that the b register is untouched so the right side is the uncomputation of the left which is different than the Add() bloq.

Reply:
the `Add` block splits the decomposition in half `_left_building_block` and `_right_building_block`. where `_right_building_block` is the uncomputation part. you can create an add bloq with the appropiate parameters and do `yield from b._left_building_block()` and `yield from b._right_building_block()`  instead of reimplementing the decomposition.

Reply:
Two problems with this:
1. b._right_building_block() isn't the same as the uncomputation of b._left_building_block() because it is in place addition (one of the CX gates is on the wrong qubit pair for full uncomputation
2. cirq.CX() does not support classical simulation and I think we want this for testing purposes

Reply:
> cirq.CX() does not support classical simulation and I think we want this for testing purposes

The solution to this should be to replace the usages of `cirq.CX` with `CNOT().on_registers()` because the `Bloq` API now supports `on_registers` to yield a cirq operation. Cirq style decompositions should continue to be supported as first class citizens in the library.

Reply:
The CX will be removed in #https://github.com/quantumlib/Qualtran/pull/585 fyi

Reply:
After https://github.com/quantumlib/Qualtran/pull/585 is merged I will modify GreaterThan to use add.left_building_block(). Is it okay to keep the right side as-is because the uncomputation operation is different or do you have any suggestions on shortening the right uncomputation?

Reply:
I need to look more carefully, but from the comments, is the issue that there is a bug in the Add right_building_block or just that it is necessarily different?

Reply:
It is necessarily different. At the end of Add(x, y): y=x+y. For GreaterThan(x,y,target) y needs to be the same so the uncomputation differs by the control/target of one CNOT.

Reply:
Right, that makes sense then.

Reply:
I don't want to block this PR but I think we should revisit this and do some refactoring to avoid having two slightly different variations of the same decomposition, but that can be a follow up.

a larger question is about the different decomposition of the same bloq. this decomposition uses 4n Ts but is linear in depth. the older estimate of 8n is based on what we did for the LessThanEqual bloq and at that time we prefered the logarithmic depth decomposition (8n Ts, log(n) depth) over the (4n Ts, n depth). I keep trying to put a proposal to track these but nothing feels good enough. anyway I will propose something asap

Reply:
Do the chemistry bloqs dependent on `GreaterThan()` prefer the logarithmic depth decomposition? If that's the case then it would make fixing the tests simpler - maintain a way to add `GreaterThan()` with logarithmic depth decomposition.

I think having both implementations is preferable anyway since it seems in the paper we want to show that Qualtran can be used to compare differing implementations of the same algorithm to compare various resource-optimized metrics.

Not sure on the best way to organize them, but if that's what we want I can just create a second `GreaterThan()` bloq with the optimal T-count decomposition and file a tracking issue to go back and combine these bloqs when an idea for bloqs with differently optimized decompositions is proposed. LMK what the preferred resolution is.

Reply:
I think we should have both implementations to maintain the current order until we find a neat way to switch between decomposition strategies (or not). Actually what you've done is a step towards making the chemistry costs more consistent with the quoted literature values (#235), currently the tests have several "corrections" to use 4n instead of 8n (the current cost), but to avoid some accounting pain it's probably best to maintain the current costs until we can replace all the comparators (rather than 2 less thans having a certain cost and a greater than having a different cost).

Reply:
I'm also happy to make the changes to the chemistry unit tests to reflect this though, if we don't want to maintain both.

Reply:
> I don't want to block this PR

yeah, I'd prefer to merge-then-organize than vice versa.

> I keep trying to put a proposal to track these

do we have an issue tracking this? I think we might. Can you update it with your latest thoughts?

> I think we should have both implementations to maintain the current order until

agree

> prefered the logarithmic depth decomposition

If you forced me to choose, I think for the regime we most care about when we have like 1 T-factory, the T _count_ is more important than the T _depth_. There's no consensus in the literature on this cost model, so we should have flags or whatever to let the user decide what they want.

Reply:
For now I created `LinearDepthGreaterThan()` separate from `GreaterThan()`.